src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -643,7 +660,163 @@ private static size_t CurlSendCallback(IntPtr buffer, size_t size, size_t nitems
+                return numBytes;
+            }
+
+            /// <summary>
+            /// Transfers up to <paramref name="length"/> data from the <paramref name="easy"/>'s
+            /// request content (non-memory) stream to the buffer.
+            /// </summary>
+            /// <returns>The number of bytes transferred.</returns>
+            private static size_t TransferDataFromRequestStream(IntPtr buffer, int length, EasyRequest easy)
+            {
+                Debug.Assert(!(easy._requestContentStream is MemoryStream), "Memory streams should use TransferFromRequestMemoryStreamToBuffer.");
+
+                MultiAgent multi = easy._associatedMultiAgent;
+
+                // First check to see whether there's any data available from a previous asynchronous read request.
+                // If there is, the transfer state's Task field will be non-null, with its Result representing
+                // the number of bytes read.  The Buffer will then contain all of that read data.  If the Count
+                // is 0, then this is the first time we're checking that Task, and so we populate the Count
+                // from that read result.  After that, we can transfer as much data remains between Offset and
+                // Count.  Multiple callbacks may pull from that one read.
+
+                EasyRequest.SendTransferState sts = easy._sendTransferState;
+                if (sts != null)
+                {
+                    // Is there a previous read that may still have data to be consumed?
+                    if (sts._task != null)
+                    {
+                        Debug.Assert(sts._task.IsCompleted, "The task must have completed if we're getting called back.");
+
+                        // Determine how many bytes were read on the last asynchronous read.
+                        // If nothing was read, then we're done and can simply return 0 to indicate
+                        // the end of the stream.
+                        int bytesRead = sts._task.GetAwaiter().GetResult(); // will throw if read failed
+                        Debug.Assert(bytesRead >= 0 && bytesRead <= sts._buffer.Length, "ReadAsync returned an invalid result length: " + bytesRead);
+                        if (bytesRead == 0)
+                        {
+                            multi.VerboseTrace("End of stream from stored task", easy: easy);
+                            sts.Update(null, 0, 0);
+                            return 0;
+                        }
+
+                        // If Count is still 0, then this is the first time after the task completed
+                        // that we're examining the data: transfer the bytesRead to the Count.
+                        if (sts._count == 0)
+                        {
+                            multi.VerboseTrace("ReadAsync completed with bytes: " + bytesRead, easy: easy);
+                            sts._count = bytesRead;
```

Comment:
The function was meant to be a simple helper for the case where all three fields needed to be updated, and I used it in all such places.  Using it in cases where only a subset need to be updated make either the helper more complex (dealing with optional arguments) or the call sites more complex (passing in the current value to overwrite the current value).  I'm going to leave this as is for now.